halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+        let selectors = std::mem::take(&mut assembly.selectors);
+        cs.directly_convert_selectors_to_fixed(selectors)
+    };
+    let mut fixed: Vec<_> = fixed.into_iter().map(|p| p.values).collect();
+    fixed.extend(selector_polys.into_iter());
+
+    let cs2 = ConstraintSystemV2Backend {
+        num_fixed_columns: cs.num_fixed_columns,
+        num_advice_columns: cs.num_advice_columns,
+        num_instance_columns: cs.num_instance_columns,
+        num_challenges: cs.num_challenges,
+        unblinded_advice_columns: cs.unblinded_advice_columns.clone(),
+        advice_column_phase: cs.advice_column_phase.iter().map(|p| p.0).collect(),
+        challenge_phase: cs.challenge_phase.iter().map(|p| p.0).collect(),
+        // TODO: Clean up all the Expression -> Expression conversions
+        gates: cs
+            .gates
+            .iter()
+            .map(|g| GateV2Backend {
+                name: g.name.clone(),
+                constraint_names: g.constraint_names.clone(),
+                polys: g.polys.clone().into_iter().map(|e| e.into()).collect(),
+            })
+            .collect(),
+        permutation: cs.permutation.clone(),
+        lookups: cs
+            .lookups
+            .iter()
+            .map(|l| lookup::ArgumentV2 {
+                name: l.name.clone(),
+                input_expressions: l
+                    .input_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+                table_expressions: l
+                    .table_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+            })
+            .collect(),
+        shuffles: cs
+            .shuffles
+            .iter()
+            .map(|s| shuffle::ArgumentV2 {
+                name: s.name.clone(),
+                input_expressions: s
+                    .input_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+                shuffle_expressions: s
+                    .shuffle_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+            })
+            .collect(),
+        general_column_annotations: cs.general_column_annotations.clone(),
+    };
+    let preprocessing = PreprocessingV2 {
+        permutation: assembly.permutation,
+        fixed,
+    };
+
+    Ok((
+        CompiledCircuitV2 {
+            cs: cs2,
+            preprocessing,
+        },
+        config,
+        cs,
+    ))
+}
+
+impl<F: Field> ConstraintSystemV2Backend<F> {
+    pub(crate) fn collect_queries(
+        &self,
+    ) -> (
+        Queries,
+        Vec<Gate<F>>,
+        Vec<lookup::Argument<F>>,
+        Vec<shuffle::Argument<F>>,
```

Comment:
Shuffle Argument collection is not really something I'd include in `collect_queries`. Meaning I'd prefer a different fn name. (Just a taste thing, nothing wrong with it in particular).

Reply:
Which name would you use?
Collect queries means: go over all expressions that will be evaluated by prover and collect the queries they do.  That includes expressions in gates, lookups and shuffles.

Reply:
Not sure TBH... Maybe we can just leave as is for now :)